### PR TITLE
[PM-3565] Enforce higher minimum KDF

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Changes minimum iterations for PBKDF2 to 600 000. Also converts the constants into ranges to ensure there is only a single place for all checks.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by Jingo88 — DISMISSED** [review]

File: libs/common/src/enums/kdf-type.enum.ts [hunk 1]
```diff
@@ -5,11 +5,23 @@ export enum KdfType {
...
+export const ARGON2_MEMORY = new RangeConstant(16, 1024, 64);
+export const ARGON2_PARALLELISM = new RangeConstant(1, 16, 4);
+export const ARGON2_ITERATIONS = new RangeConstant(1, 10, 3);
 
 export const DEFAULT_KDF_TYPE = KdfType.PBKDF2_SHA256;
-export const DEFAULT_PBKDF2_ITERATIONS = 600000;
-export const DEFAULT_KDF_CONFIG = new KdfConfig(DEFAULT_PBKDF2_ITERATIONS);
+export const PBKDF2_ITERATIONS = new RangeConstant(600_000, 2_000_000, 600_000);
+export const DEFAULT_KDF_CONFIG = new KdfConfig(PBKDF2_ITERATIONS.defaultValue);
 export const SEND_KDF_ITERATIONS = 100000;
```
Comment (aj-rosado) [inline libs/common/src/enums/kdf-type.enum.ts:27]: Any reason to not increase the Send KDF?

**Review by aj-rosado — COMMENTED** [review]
The description states that we are increasing to 100 000 but seems wrong as we are increasing to 600 000

File: libs/common/src/enums/kdf-type.enum.ts (hunk 1 above)
Comment (djsmith85) [reply libs/common/src/enums/kdf-type.enum.ts:27]: @aj-rosado: We have [PM-3384](https://bitwarden.atlassian.net/browse/PM-3384) tracking this.

**Review by aj-rosado — APPROVED** [review]

File: libs/common/src/enums/kdf-type.enum.ts [hunk 2]
```diff
@@ -5,11 +5,23 @@ export enum KdfType {
...
+  }
+
+  inRange(value: number): boolean {
+    return value >= this.min && value <= this.max;
+  }
+}
+
+export const ARGON2_MEMORY = new RangeConstant(16, 1024, 64);
+export const ARGON2_PARALLELISM = new RangeConstant(1, 16, 4);
+export const ARGON2_ITERATIONS = new RangeConstant(1, 10, 3);
```
Comment (djsmith85) [inline libs/common/src/enums/kdf-type.enum.ts:22]: The `ARGON2_ITERATIONS.min` is lower than previously set [here](https://github.com/bitwarden/clients/pull/6440/files#diff-34fe1b0da916c7483304b7b1d115a6b12653e6adaad5b94ac76e1b51a289a989L73)

**Review by djsmith85 — CHANGES_REQUESTED** [review]
@Hinton Nice work and overall a far better solution using the ranges.

Just one small nit/observation

File: libs/common/src/enums/kdf-type.enum.ts (hunk 2 above)
Comment (Hinton) [reply libs/common/src/enums/kdf-type.enum.ts:22]: Interesting. Server used 1 as the minimum value.

**Review by djsmith85 — DISMISSED** [review]

**Review by MGibson1 — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by Jingo88 — APPROVED** [review]

**Comment by Hinton** [comment]
I had to re-work the approach slightly. `makeKey` is used during login/unlock and other operations. I extracted all the validation into a separate function since it's unlikely we would want to prevent users from logging in.

**Review by MGibson1 — DISMISSED** [review]

**Review by djsmith85 — DISMISSED** [review]

**Review by Jingo88 — APPROVED** [review]

**Review by djsmith85 — DISMISSED** [review]

**Review by djsmith85 — DISMISSED** [review]

**Review by MGibson1 — APPROVED** [review]

**Review by Jingo88 — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by MGibson1 — DISMISSED** [review]

**Review by Jingo88 — DISMISSED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by Jingo88 — APPROVED** [review]

**Review by jlf0dev — APPROVED** [review]

**Pull request merged**